### ISSUE TITLE
Use AxisArrays

### ISSUE THREAD

**Issue by trappmartin** [TuringLang/MCMCChains.jl#20]
Currently, the Chains type has a problem specific implementation which can be done using AxisArrays. Using AxisArrays as a field of Chains would simplify the code base, e.g. getindex and setindex.

The following minimal example illustrates how the `Chains` type could look like if we use AxisArrays.

```julia
struct Chains{T<:Real}
    value::AxisArray
    logevidence::T
end

function Chains(val::AbstractArray{T,3}; ...) where T<:Real
    names = [:iter, :var, :chain]
    axvals = [
              1:size(val, 1),
              map(i -> Symbol("Param$i"), 1:size(val, 2)),
              map(i -> Symbol("Chain$i"), 1:size(val, 3))
    ]
    
    axs = ntuple(i -> Axis{names[i]}(axvals[i]), 3)
    A = AxisArray(convert(Array{Union{Missing,T},3}, val), axs...)
    return Chains{T}(A, zero(T))
end

Base.getindex(c::Chains, i...) = getindex(c.values, i...)
Base.setindex!(c::Chains, v, i...) = setindex!(c.values, v, i...)
```
This would allow us to index an instance of `Chains` as follows:
```julia
chn = Chains(randn(100, 5, 2));
chn[:,"Param1",:] # Sample values for Param1 at all iterations and chains.
chn[:,1,:] # Sample values for Param1 at all iterations and chains.
chn[Axis{:iter}(1 .. 50)] # Sample values of all parameters at all chains for iterations 1 to 50
```

**Comment by trappmartin** [comment]
Related issue: #23

**Comment by fisherc2** [comment]
Hi trappmartin. I came across this issue when I discovered that I cannot subset on chains by parameter name. Out of curiosity, would your proposal allow the following behavior:

```
subchain = chain[:,["parm1","parm2"],:]
describe(subchain)
```
I find this functionality to be useful for plotting and post analyses.

**Comment by goedman** [comment]
Hi Chris, Cameron ( @cpfiffer ) is also looking at the chain object ( https://github.com/TuringLang/MCMCChain.jl/issues/29#issuecomment-458220639 ).

**Comment by cpfiffer** [comment]
Closed by #33.
